wire-ios-data-model/Source/MLS/MLSService.swift

```diff
@@ -1250,18 +1250,48 @@ public final class MLSService: MLSServiceInterface {
             try ZMConversation.fetchConversationsWithMLSGroupStatus(
                 mlsGroupStatus: .pendingJoin,
                 in: context
-            ).compactMap(\.mlsGroupID)
+            )
         }
 
         logger.info("joining \(pendingGroups.count) group(s)")
 
         await withTaskGroup(of: Void.self) { group in
             for pendingGroup in pendingGroups {
                 group.addTask {
+                    guard let mlsGroupID = await context.perform({ pendingGroup.mlsGroupID }) else {
+                        return
+                    }
+
                     do {
-                        try await self.joinByExternalCommit(groupID: pendingGroup)
+                        let (epoch, isSelfConversation) = await context.perform {
+                            (pendingGroup.epoch, pendingGroup.isSelfConversation)
+                        }
+                        let conversationExists = try await self.conversationExists(
+                            groupID: mlsGroupID
+                        )
+                        let shouldEstablishGroup = epoch == 0 && isSelfConversation && !conversationExists
+
+                        if shouldEstablishGroup {
+
+                            let mlsUsers = await context.perform {
+                                pendingGroup.localParticipants.map(MLSUser.init)
+                            }
+
+                            let ciphersuite = try await self.establishGroup(
+                                for: mlsGroupID,
+                                with: mlsUsers
+                            )
+
+                            await context.perform {
+                                pendingGroup.ciphersuite = ciphersuite
+                                pendingGroup.mlsStatus = .ready
+                            }
+
+                        } else {
+                            try await self.joinByExternalCommit(groupID: mlsGroupID)
+                        }
                     } catch {
-                        WireLogger.mls.error("Failed to join pending group (\(pendingGroup): \(error)")
+                        WireLogger.mls.error("Failed to join pending group (\(mlsGroupID): \(error)")
```

Comment:
suggestion: the mlsGroupId should be obfuscated.
```suggestion
                        WireLogger.mls.error("Failed to join pending group: \(error)", attributes: [.mlsGroupID: mlsGroupID.safeForLoggingDescription])
```